### DISCUSSION TITLE
Tuple Syntax for routing

### DISCUSSION THREAD

**Discussion by arthur-er** [adonisjs/core#3743] [Ideas]
This discussion started on [Twitter](https://twitter.com/ProjektGopher/status/1530928862258835456)

As pointed by Len Woodward on the starting discussion, the advantages of the tuple syntax are go-to click on IDE's and enforced existing methods on controllers (leveraging Typescript).

And as pointed by @thetutlage the main disadvantages are a cluttered routes file, and an increased boot time.

For me:
- The increased boot time only matters in development, as in production the app only boots once per restart
- The visual noise that the imports bring don't matter as the majority of time I'm not looking at the start of the routing file, but at the route I'm defining
- The click to definition would be nice (although I think the AdonisJS extension already handles this in vs code, but I'm not using it atm)
- The increased type security would also be nice

And in JS/TS land, the Class.name syntax that would cause more noise, is not needed, as we can use just Class

**Comment by RomainLanz** [comment]
> The visual noise that the imports bring don't matter as the majority of time I'm not looking at the start of the routing file, but at the route I'm defining

This is not a real concern IMO. Most IDE folds the imports when you have too many.

**Reply by davidharting** [reply]
In some ways, the imports are an upside as you are getting some compile time safety that you are referencing a real controller. 

Although I believe I recall reading something about type safety eventually coming for the controller strings as well.

**Reply by thetutlage** [reply]
> Although I believe I recall reading something about type safety eventually coming for the controller strings as well.

That's my wish from TypeScript team as there are open issues that adheres to their design principle. But, I have no insights if/when they will add that feature.

I think, we can optionally support that tuple syntax for the one's who wants to use it. My main concern has always been that pulling controllers within routes will actually import in the entire codebase during the boot cycle of the app and will impact the performance.

**Comment by ProjektGopher** [comment]
Oh hey, it's my twitter reply 😄 

Thx for starting the conversation @arthur-er, I'll be interested to see what happens here 👀
